bundler/lib/bundler/checksum.rb

```diff
@@ -0,0 +1,228 @@
+# frozen_string_literal: true
+
+module Bundler
+  class Checksum
+    DEFAULT_ALGORITHM = "sha256"
+    DEFAULT_BLOCK_SIZE = 16_384
+    private_constant :DEFAULT_BLOCK_SIZE
+
+    class << self
+      def from_gem(io, pathname, algo = DEFAULT_ALGORITHM)
+        digest = Bundler::SharedHelpers.digest(algo.upcase).new
+        digest << io.read(DEFAULT_BLOCK_SIZE) until io.eof?
+        Checksum.new(algo, digest.hexdigest!, Source.new(:gem, pathname))
+      end
+
+      def from_api(digest, source_uri)
+        # transform the bytes from base64 to hex
+        hexdigest = digest.length == 44 ? digest.unpack("m0").first.unpack("H*").first : digest
```

Comment:
note: add a comment to use `unpack1` once we drop older rubies, it avoids the array allocation